src/Hl7.Fhir.Base/Serialization/PathStack.cs

```diff
@@ -43,33 +51,68 @@ public void ExitResource()
                 throw new InvalidOperationException("Cannot exit a resource while inside a property.");
             }
             else
+            {
                 _path = _path.Substring(0, top);
+            }
+
+            if (_paths.Count() == 1)
+                _paths.Pop();
         }
 
-        public void EnterElement(string name)
+        public void EnterElement(string name, int? index, bool isPrimitive)
         {
             bool empty = !_tops.Any();
             _tops.Push(_path.Length);
+            _indexer.Push(index ?? 0);
             _path += empty ? name : $".{name}";
+            if (!isPrimitive)
+            {
+                if (index.HasValue)
```

Comment:
The original stack was pretty conservative in memory usage since we had 1 string + a stack of integer indices. The new stack now has those PLUS a stack of strings. That's a clear step backwards.  Couldn't we still build a single string (with the index markers) and have, say, two stacks (or a stack with valuetuples) with indices so we can construct both the definition and the instance path from the same string?

Reply:
We discussed this during the call - there are no new strings allocated, so it should be fine.